routing/payment_state_machine.go

```diff
@@ -0,0 +1,419 @@
+	paymentID uint64, route *Route) (paymentState, error) {
+
+	// Serialize the information before opening the db transaction.
+	var b bytes.Buffer
+	if err := channeldb.WriteElements(&b, paymentID); err != nil {
+		return 0, err
+	}
+
+	if err := serializeRoute(&b, route); err != nil {
+		return 0, err
+	}
+
+	nextState := paymentStateAttemptSent
+	err := s.db.Batch(func(tx *bbolt.Tx) error {
+		bucket, err := s.getWriteBucket(tx, pHash)
+		if err != nil {
+			return err
+		}
+
+		err = s.putPaymentState(bucket, nextState)
+		if err != nil {
+			return err
+		}
+
+		// We put the serialized information under the
+		// paymentAttemptKey, such that it can be fetched when a result
+		// for this payment attempt comes back.
+		return bucket.Put(paymentAttemptKey, b.Bytes())
+	})
+	if err != nil {
+		return 0, err
+	}
+
+	return nextState, nil
+}
+
+// setStateIntermediateFailure records the given error and deletes the current
+// paymentAttempt from the DB. When this method returns a new paymentAttempt
+// can be created and sent to the Switch.
+//
+// When this method returns the payment will be in the state
+// paymentStateIntended.
+func (s *paymentStateMachine) setStateIntermediateFailure(pHash lntypes.Hash,
+	paymentErr error) (paymentState, error) {
+
+	// We store the error as a string.
+	errMsg := []byte(paymentErr.Error())
+
+	// Go back to initial state, to allow new payment attempts to be made.
+	nextState := paymentStateIntended
+	err := s.db.Batch(func(tx *bbolt.Tx) error {
+		bucket, err := s.getWriteBucket(tx, pHash)
+		if err != nil {
+			return err
+		}
+
+		err = s.putPaymentState(bucket, nextState)
+		if err != nil {
+			return err
+		}
+
+		// Since a result for the payment attempt came back, and we are
+		// moving back to the paymentStateIntended state, we can safely
+		// remove the previous paymentAttempt information.
+		if err := bucket.Delete(paymentAttemptKey); err != nil {
+			return err
+		}
+
+		// Store the error for record keeping.
+		return bucket.Put(paymentErrorKey, errMsg)
+	})
+	if err != nil {
+		return 0, err
+	}
+
+	return nextState, nil
+}
+
+// setStateTerminalFailure records the given error and deletes the current
+// paymentAttempt from the DB. A payment in this state is safe to attempt to
+// resend. It is kept in the database in this state to be available for
+// historical lookups.
+//
+// When this method returns the payment will be in the state
+// paymentStateFailed.
+func (s *paymentStateMachine) setStateTerminalFailure(pHash lntypes.Hash,
+	paymentErr error) (paymentState, error) {
+
+	// We store the error as a string.
+	errMsg := []byte(paymentErr.Error())
+
+	// Go to the failed state, where new payment attempts are allowed.
+	nextState := paymentStateFailed
+	err := s.db.Batch(func(tx *bbolt.Tx) error {
+		bucket, err := s.getWriteBucket(tx, pHash)
+		if err != nil {
+			return err
+		}
+
+		err = s.putPaymentState(bucket, nextState)
```

Comment:
The body of this method can be consolidated with `setStateIntermediateFailure`.

Reply:
Fixed: https://github.com/lightningnetwork/lnd/pull/2761/commits/b45ef5b0684ed8b4ecf9db81ab34ec604f6fcded